src/vmspawn/vmspawn-util.c

```diff
@@ -0,0 +1,341 @@
+        assert(confs || n_confs == 0);
+
+        FOREACH_ARRAY(conf, confs, n_confs)
+                free(conf->name);
+
+        free(confs);
+}
+
+static int ovmf_config_file_compare(const OvmfConfigFile *a, const OvmfConfigFile *b) {
+        int name_cmp = strcmp(a->name, b->name);
+        return (name_cmp != 0) ? name_cmp : CMP(a->priority, b->priority);
+}
+
+/* holds the data retrieved from the QEMU firmware interop JSON data */
+typedef struct FirmwareData {
+        char **features;
+        char *firmware;
+        char *vars;
+} FirmwareData;
+
+static FirmwareData* firmware_data_new(void) {
+        return new0(FirmwareData, 1);
+}
+
+static FirmwareData* firmware_data_free(FirmwareData *fwd) {
+        fwd->features = strv_free(fwd->features);
+        fwd->firmware = mfree(fwd->firmware);
+        fwd->vars = mfree(fwd->vars);
+
+        return mfree(fwd);
+}
+DEFINE_TRIVIAL_CLEANUP_FUNC(FirmwareData*, firmware_data_free);
+
+static int firmware_executable(const char *name, JsonVariant *v, JsonDispatchFlags flags, void *userdata) {
+        JsonDispatch table[] = {
+                { "filename", JSON_VARIANT_STRING, json_dispatch_string, offsetof(FirmwareData, firmware), JSON_MANDATORY },
+                { "format",   JSON_VARIANT_STRING, NULL,                 0,                                JSON_MANDATORY },
+                {}
+        };
+
+        return json_dispatch(v, table, NULL, 0, userdata);
+}
+
+static int firmware_nvram_template(const char *name, JsonVariant *v, JsonDispatchFlags flags, void *userdata) {
+        JsonDispatch table[] = {
+                { "filename", JSON_VARIANT_STRING, json_dispatch_string, offsetof(FirmwareData, vars), JSON_MANDATORY },
+                { "format",   JSON_VARIANT_STRING, NULL,                 0,                            JSON_MANDATORY },
+                {}
+        };
+
+        return json_dispatch(v, table, NULL, 0, userdata);
+}
+
+static int firmware_mapping(const char *name, JsonVariant *v, JsonDispatchFlags flags, void *userdata) {
+        JsonDispatch table[] = {
+                { "device",         JSON_VARIANT_STRING, NULL,                    0, JSON_MANDATORY },
+                { "executable",     JSON_VARIANT_OBJECT, firmware_executable,     0, JSON_MANDATORY },
+                { "nvram-template", JSON_VARIANT_OBJECT, firmware_nvram_template, 0, JSON_MANDATORY },
+                {}
+        };
+
+        return json_dispatch(v, table, NULL, 0, userdata);
+}
+
+int find_ovmf_config(int search_sb, OvmfConfig** ret) {
+        _cleanup_(ovmf_config_freep) OvmfConfig* config = NULL;
+        int r;
+
+        config = new0(OvmfConfig, 1);
+        if (!config)
+                return -ENOMEM;
+
+        _cleanup_free_ char *xdg_config_dir = NULL, *user_firmware_dir;
+        r = xdg_user_config_dir(&xdg_config_dir, NULL);
+        if (r < 0)
+                return r;
+
+        user_firmware_dir = strjoin(xdg_config_dir, "/qemu/firmware");
```

Comment:
use the 2nd param of xdg_user_config_dir() instead?

Reply:
Fixed in 69a7728f8a.